### PR TITLE
feat: Badges implementation

### PR DESCRIPTION
**Pull request by sandrade-dcl**
## What does this PR change?
This PR integrates Badges in Explorer Alpha.

**Documentation**: https://www.notion.so/decentraland/Badges-Implementation-274e8b01cef44353b34119423aa1dd5d
**Figma**: https://www.figma.com/design/6DtX7mRUlK9UP2AcRp1obW/%F0%9F%97%BA%EF%B8%8F-Badges-%7C-Explorer-%7C-07%2F24?node-id=7-1126&t=9J59kndjIEyTATPG-0

image

## How to test the changes?
1. Launch the explorer.
2. Test Badges feature.

## Our Code Review Standards
https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

## Summary by CodeRabbit

- **New Features**
  - Enhanced the passport management system with improved navigation between sections, allowing users to directly access OVERVIEW or BADGES.
  - Introduced new methods for managing the visibility of passport sections based on user interactions, improving overall user experience.
  - Integrated badge notifications to allow users to open the passport directly from notifications.
  - Expanded UI capabilities with new panels for badges and overview sections, enhancing user interaction.

- **Bug Fixes**
  - Optimized internal state management to prevent unnecessary reloading of passport sections, enhancing performance.

- **Style**
  - Improved UI design for better visibility and interaction with different passport sections, including layout adjustments for various screen resolutions.

### CODE REVIEW

File: Explorer/Assets/DCL/BadgesAPIService/LatestAchievedBadgesResponse.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+    [Serializable]
+    public class LatestAchievedBadgeData
+    {
+        public string id;
+        public string name;
+        public string tierName;
+        public string image;
+    }
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/BadgesAPIService/LatestAchievedBadgesResponse.cs:25]: Firstly you suggest to convert the fields into properties and after it you suggest to make them `readonly`, but a property cannot be `readonly`. What are you suggesting then?

**Review by mikhail-dcl — CHANGES_REQUESTED** [review]
I propose 2 areas for improvement:
- Memory allocations, profile the passport and see how much is allocated every time. I mentioned my concerns regarding `APIClient` but there could be an issue on the view level as well
- Code clarity in SubControllers and their views: please review redundancies, try to organize null checks without repetitions and with some contextual names so it will be clear why and for what purpose they are executed.

File: Explorer/Assets/DCL/BadgesAPIService/BadgesAPIClient.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            foreach (var badge in badgesResponse.data.achieved)
+                badgesInfo.achieved.Add(ResponseToBadgeInfo(badge, false));
+
+            foreach (var badge in badgesResponse.data.notAchieved)
+                badgesInfo.notAchieved.Add(ResponseToBadgeInfo(badge, true));
+
+            return badgesInfo;
+        }
+
+        private static BadgeInfo ResponseToBadgeInfo(BadgeData badge, bool isLocked) =>
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/BadgesAPIService/BadgesAPIClient.cs:79]: Done: https://github.com/decentraland/unity-explorer/pull/1718/commits/bb68712aaab9e1252bf984d2399ff97cdfa7ffe5

File: Explorer/Assets/DCL/Notifications/NewNotification/NewNotificationController.cs [hunk 3]
```diff
@@ -119,6 +127,22 @@ private async UniTask ProcessDefaultNotificationAsync(INotification notification
             await AnimateNotificationCanvasGroupAsync(viewInstance.NotificationViewCanvasGroup);
         }
 
+        private async UniTask ProcessBadgeNotificationAsync(INotification notification)
+        {
+            if (viewInstance == null)
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Notifications/NewNotification/NewNotificationController.cs]: Done: https://github.com/decentraland/unity-explorer/pull/1718/commits/02e063230f0e28a7948b839884415d2047444006

File: Explorer/Assets/DCL/NotificationsBusController/NotificationsBus/NotificationsBusController.cs [hunk 4]
```diff
@@ -36,5 +37,27 @@ public void SubscribeToNotificationTypeReceived(NotificationType desiredType, No
             thisEvent += listener;
             notificationReceivedSubscribers[desiredType] = thisEvent;
         }
+
+        public void SubscribeToAllNotificationTypesClick(NotificationClickedDelegate listener)
+        {
+            for (var i = 0; i < Enum.GetValues(typeof(NotificationType)).Length; i++)
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/NotificationsBusController/NotificationsBus/NotificationsBusController.cs]: Done: https://github.com/decentraland/unity-explorer/pull/1718/commits/b131db7fd4ac022343ec89eb9a08531ec039d6c8

File: Explorer/Assets/DCL/Passport/Fields/Badges/BadgeDetailCard_PassportFieldView.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        public void SetAsSelected(bool isSelected)
+        {
+            SelectedOutline.SetActive(isSelected);
+            IsSelected = isSelected;
+        }
+
+        public void SetAsNew(bool isNew) =>
+            NewMark.SetActive(isNew);
+
+        public void Setup(BadgeInfo badgeInfo, bool isOwnProfile)
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Passport/Fields/Badges/BadgeDetailCard_PassportFieldView.cs:94]: Completely agreed.
I have splitted the logic in several descriptive methods.
Done here: https://github.com/decentraland/unity-explorer/pull/1718/commits/a245dc1a6a8dbbd6990c4495f1e737b0bde9ff4b

File: Explorer/Assets/DCL/Passport/Modules/Badges/BadgeDetailsCards_PassportModuleSubController.cs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                badge.isNew = BadgesUtils.IsBadgeNew(badge.id);
+
+            var badgeDetailCard = badgeDetailCardsPool.Get();
+            badgeDetailCard.Setup(badge, isOwnProfile);
+            badgeDetailCard.Button.onClick.AddListener(() => { SelectBadgeCard(badgeDetailCard, isOwnProfile); });
+
+            // Place badge into the corresponding category container
+            foreach (var badgesCategoryContainer in badgesCategoriesController.InstantiatedBadgesCategoryContainers)
+            {
+                if (!string.Equals(badgesCategoryContainer.Category, badge.category, StringComparison.CurrentCultureIgnoreCase))
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Passport/Modules/Badges/BadgeDetailsCards_PassportModuleSubController.cs]: Done: https://github.com/decentraland/unity-explorer/pull/1718/commits/426612af3c24532d0883d3e8c704d1b314af2b00

File: Explorer/Assets/DCL/Passport/Modules/Badges/BadgesCategories_PassportModuleSubController.cs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+using Object = UnityEngine.Object;
+
+namespace DCL.Passport.Modules.Badges
+{
+    public class BadgesCategories_PassportModuleSubController
+    {
+        private const string ALL_FILTER = "All";
+        private const int BADGES_CATEGORIES_POOL_DEFAULT_CAPACITY = 6;
+
+        public readonly List<ButtonWithSelectableStateView> InstantiatedBadgesFilterButtons = new ();
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Passport/Modules/Badges/BadgesCategories_PassportModuleSubController.cs:15]: They are already `readonly` 👍

File: Explorer/Assets/DCL/Passport/Modules/Badges/BadgeInfo_PassportModuleSubController.cs [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+                }
+            }
+        }
+
+        private void SelectBadgeTier(int tierIndex, BadgeInfo badgeInfo)
+        {
+            var tier = currentTiers[tierIndex];
+            badgeInfoModuleView.BadgeNameText.text = $"{badgeInfo.name} {tier.tierName}";
+            string tierCompletedAt = badgeInfo.GetTierCompletedDate(tier.tierId);
+            badgeInfoModuleView.BadgeDateText.text = !string.IsNullOrEmpty(tierCompletedAt) ? $"Unlocked: {BadgesUtils.FormatTimestampDate(tierCompletedAt)}" : "Locked";
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Passport/Modules/Badges/BadgeInfo_PassportModuleSubController.cs]: Simplified: https://github.com/decentraland/unity-explorer/pull/1718/commits/5b59af3026acaeaf6fa35f54f188fb1af2e9d090

File: Explorer/Assets/DCL/Passport/Modules/Badges/BadgeInfo_PassportModuleSubController.cs [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+            badgeInfoModuleView.Badge3DImage.color = badgeInfo.isLocked ? badgeInfoModuleView.Badge3DImageLockedColor : badgeInfoModuleView.Badge3DImageUnlockedColor;
+            badgeInfoModuleView.Badge3DAnimator.SetBool(IS_STOPPED_3D_IMAGE_ANIMATION_PARAM, badgeInfo.isLocked);
+
+            if (!badgeInfo.isTier)
+            {
+                badgeInfoModuleView.BadgeNameText.text = badgeInfo.name;
+                badgeInfoModuleView.BadgeDateText.text = !badgeInfo.isLocked ? $"Unlocked: {BadgesUtils.FormatTimestampDate(badgeInfo.completedAt)}" : "Locked";
+                badgeInfoModuleView.BadgeDescriptionText.text = badgeInfo.description;
+                int simpleBadgeProgressPercentage = badgeInfo.progress.stepsDone * 100 / badgeInfo.progress.totalStepsTarget;
+                badgeInfoModuleView.SimpleBadgeProgressBarFill.sizeDelta = new Vector2(simpleBadgeProgressPercentage * (badgeInfoModuleView.SimpleBadgeProgressBar.sizeDelta.x / 100), badgeInfoModuleView.SimpleBadgeProgressBarFill.sizeDelta.y);
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Passport/Modules/Badges/BadgeInfo_PassportModuleSubController.cs]: More simplifications: https://github.com/decentraland/unity-explorer/pull/1718/commits/28377ea9d5816db5e010c5ce74a4dec8a3e14fa2

File: Explorer/Assets/DCL/Passport/Modules/Badges/BadgeInfo_PassportModuleSubController.cs [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+                {
+                    SelectBadgeTier(i, currentBadgeInfo);
+
+                    loadBadge3DImageCts = loadBadge3DImageCts.SafeRestart();
+                    LoadBadge3DImageAsync(tierButton.Model.assets, loadBadge3DImageCts.Token).Forget();
+                }
+            }
+        }
+
+        private void SelectBadgeTier(int tierIndex, BadgeInfo badgeInfo)
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Passport/Modules/Badges/BadgeInfo_PassportModuleSubController.cs:199]: I don't know what do you mean, the `BadgeInfo_PassportModuleView` view is completely empty of logic.

File: Explorer/Assets/DCL/Passport/Modules/Badges/BadgeInfo_PassportModuleSubController.cs [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+
+                if (assets?.textures3d == null)
+                    return;
+
+                string baseColorUrl = assets.textures3d.baseColor;
+                string normalUrl = assets.textures3d.normal;
+                string hrmUrl = assets.textures3d.hrm;
+
+                Texture2D baseColorTexture = await webRequestController.GetTextureAsync(new CommonArguments(URLAddress.FromString(baseColorUrl)), new GetTextureArguments(false), GetTextureWebRequest.CreateTexture(TextureWrapMode.Clamp), ct);
+                baseColorTexture.filterMode = FilterMode.Bilinear;
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Passport/Modules/Badges/BadgeInfo_PassportModuleSubController.cs]: That's true! Done: https://github.com/decentraland/unity-explorer/pull/1718/commits/e3eb1b259c6e49854e475eba397a268479050c40

File: Explorer/Assets/DCL/Passport/PassportController.cs [hunk 12]
```diff
@@ -133,46 +163,70 @@ protected override void OnViewClose()
...
         }
 
-        private async UniTaskVoid LoadUserProfileAsync(string userId, CancellationToken ct)
+        private async UniTaskVoid LoadPassportSectionAsync(string userId, PassportSection sectionToLoad, CancellationToken ct, string? badgeIdSelected = null)
         {
             try
             {
+                switch (sectionToLoad)
+                {
+                    case PassportSection.OVERVIEW when overviewSectionAlreadyLoaded:
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Passport/PassportController.cs]: Good idea!
Done here: https://github.com/decentraland/unity-explorer/pull/1718/commits/f564a060cbdec7a9947be75419bc2e2b65deb321

File: Explorer/Assets/DCL/BadgesAPIService/BadgesAPIClient.cs [hunk 13]
```diff
@@ -0,0 +1,94 @@
+using Cysharp.Threading.Tasks;
+using DCL.Diagnostics;
+using DCL.Multiplayer.Connections.DecentralandUrls;
+using DCL.WebRequests;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace DCL.BadgesAPIService
+{
+    public class BadgesAPIClient
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/BadgesAPIService/BadgesAPIClient.cs:11]: Great suggestion! I've applied it and now all the lists managed by the api client are being pooled/released 👍 

Done here: https://github.com/decentraland/unity-explorer/pull/1718/commits/a9eedafbffa7cd05d50e0e374562753b4ea8d087

File: Explorer/Assets/DCL/Passport/PassportController.cs [hunk 14]
```diff
@@ -183,10 +237,85 @@ private async UniTaskVoid LoadUserProfileAsync(string userId, CancellationToken
...
+        {
+            string badgeIdToOpen = string.Empty;
+            if (parameters.Length > 0 && parameters[0] is BadgeGrantedNotification badgeNotification)
+                badgeIdToOpen = badgeNotification.Metadata.Id;
+
+            openPassportFromBadgeNotificationCts = openPassportFromBadgeNotificationCts.SafeRestart();
+            OpenPassportFromBadgeNotificationAsync(badgeIdToOpen, openPassportFromBadgeNotificationCts.Token).Forget();
+        }
+
+        private async UniTaskVoid OpenPassportFromBadgeNotificationAsync(string badgeIdToOpen, CancellationToken ct)
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Passport/PassportController.cs:307]: Done: https://github.com/decentraland/unity-explorer/pull/1718/commits/22d134424bd2f24030dec26e8940509ba4209847

**Review by mikhail-dcl — APPROVED** [review]

**Pull request merged**